### ISSUE TITLE
Support for Microsoft Graph API

### ISSUE THREAD

**Issue by grtn316** [hashicorp/terraform-provider-azurerm#12443]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Support ends for Azure AD Graph APIs 6/30/2022 and new features were discontinued 6/30/2020 ([link](https://techcommunity.microsoft.com/t5/azure-active-directory-identity/update-your-applications-to-use-microsoft-authentication-library/ba-p/1257363)).

It appears that the TF provider is still using the Azure AD Graph APIs to authenticate SPNs. This can be seen by inspecting traffic and using this [article](https://docs.microsoft.com/en-us/graph/migrate-azure-ad-graph-request-differences) to compare.

When will this be upgraded to the Microsoft Graph APIs?

### New or Affected Resource(s)

* latest azurerm provider

### References

* https://docs.microsoft.com/en-us/graph/migrate-azure-ad-graph-request-differences
* https://techcommunity.microsoft.com/t5/azure-active-directory-identity/update-your-applications-to-use-microsoft-authentication-library/ba-p/1257363
* https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-migration
* https://techcommunity.microsoft.com/t5/azure-active-directory-identity/have-you-updated-your-applications-to-use-the-microsoft/ba-p/1144698

**Comment by manicminer** [comment]
Hi @grtn316, thanks for noting this. This is on our roadmap and we're planning to switch to using Microsoft Graph for all AAD operations in v3.0, since this will be a breaking change for users due to the difference in permissions which will be needed.

**Comment by grtn316** [comment]
> Hi @grtn316, thanks for noting this. This is on our roadmap and we're planning to switch to using Microsoft Graph for all AAD operations in v3.0, since this will be a breaking change for users due to the difference in permissions which will be needed.

Thanks for the update. Do you happen to know how far out v3.0 is?

**Comment by manicminer** [comment]
@grtn316 We don't have a timeframe at this stage as we're still early in planning/scoping.

**Comment by dohnto** [comment]
I upgraded my az to 2.30.0 which introduced MSAL related [breaking change](https://techcommunity.microsoft.com/t5/azure-tools/azure-cli-migration-from-adal-to-msal/ba-p/2909427) and my terraform (azurerm=2.83.0) stopped working with
```
│ Error: building account: getting authenticated object ID: Error parsing json result from the Azure CLI: Error waiting for the Azure CLI: exit status 1: Could not retrieve token from local cache. Please run 'az login'.
│ 
│   with provider["registry.terraform.io/hashicorp/azurerm"],
│   on main.tf line 8, in provider "azurerm":
│    8: provider "azurerm" {
```

Is this related?

**Comment by manicminer** [comment]
@dohnto I've been doing some testing with the new az-cli version 2.30.0 and it's working great for me with single-tenant auth.

However, I've noticed that it's likely to fail when using mult-tenant auth (i.e. specifying `auxiliary_tenant_ids`) - are you perhaps using that? I believe we'll be able to get a fix out for this soon.

**Comment by DaveOHenry** [comment]
We are using multiple tenants via provider `alias` and it's not working anymore with az-cli 2.30. The error message is the same as above.

**Comment by manicminer** [comment]
@DaveOHenry That looks like an az-cli error, I'd suggest removing your `~/.azure` directory and signing in again.

If you are explicitly using multi-tenant auth (e.g. via `auxiliary_tenant_ids` / `ARM_AUXILIARY_TENANT_IDS`), there a known issue for which we have a fix incoming.

**Comment by DaveOHenry** [comment]
Thanks for the insights and pointing me into the right direction! The commands are indeed visible in the azure-cli logs. There are many `az account get-access-token` commands and one of these fails with "Errno 2".
Sorry for hijacking this issue. There is clearly a bug in the azure-cli that was most likely introduced in v2.30: https://github.com/Azure/azure-cli/issues/20252

**Comment by manicminer** [comment]
We'll be switching away from the Azure Active Directory Graph API in version 3.0 of the provider. In the meantime, this capability is available in version 2.94.0 or later of the provider via the [`use_msal` provider property](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs#use_msal).

With the current implementation, in most cases the provider will likely _not_ make any calls to Microsoft Graph; instead it will attempt to consume the claims from access tokens, and will only call Microsoft Graph as a fallback in the event that more information about the authenticated principal is needed. Either way, no additional API permissions should be required and the net effect from enabling this property should simply be that the provider ceases to call the Azure Active Directory Graph API.

**Comment by tombuildsstuff** [comment]
It's worth calling that [the AzureRM Terraform Backend](https://www.terraform.io/language/settings/backends/azurerm) will also use the Azure Active Directory Graph API by default, in Terraform 1.1 this can be switched to using Microsoft Graph by setting `use_microsoft_graph = true` - but this default will change in Terraform 1.2 to be on by default
